### DISCUSSION TITLE
HTML API: Plans for 6.5

### DISCUSSION THREAD

**Discussion by dmsnell** [WordPress/gutenberg#54579] [HTML API]
[HTML issues](https://github.com/WordPress/gutenberg/discussions/57560) | [Refactors](https://github.com/WordPress/gutenberg/discussions/54750) | [↑ Broader Roadmap](https://github.com/WordPress/gutenberg/discussions/54583) | [Plans for 6.6 →](https://github.com/WordPress/gutenberg/discussions/57827)

### Bug fixes and quality

 - [x] WordPress/gutenberg#57216
    - ~not in `trunk`/`trunk` but may affect older version of the plugin. need to figure out what to resolve, if anything~.
    - appears to be resolved already by means of the patch in Gutenberg, but that patch wasn't in the tested release candidate.
 - [x] WordPress/wordpress-develop#5793
 - [ ] Review failing tests in the html5lib and see if we can eliminate any of them.
 - [ ] Generate a function that produces the output expected by the html5lib tests.

### Continue adding support for more HTML tags

 - [x] WordPress/wordpress-develop#5535 (#57022)
 - [x] WordPress/wordpress-develop#5539
    - :shipit: Approved and ready to merge.
    - ~Currently undergoing testing work. Some test is needed for the following rule, as it may be broken somehow: "If node is in the special category, but is not an address, div, or p element, then jump to the step labeled done below."~
 - [x] WordPress/wordpress-develop#5762 (@sirreal)
    - :shipit: Approved and ready to merge.

### Add functionality to the HTML API

 - [x] WordPress/wordpress-develop#5725
 - [x] WordPress/wordpress-develop#5683

### Design work and framework improvements

 - [ ] Incorporate an nth-child counter for breadcrumbs for uniquely identifying a node in the document, for future use with CSS selectors.
 - [ ] https://github.com/WordPress/wordpress-develop/pull/5794
 - [ ] <s>Create `WP_HTML::render()` using "funky comments" to generate HTML markup from PHP.</s>
    - dmsnell/wordpress-develop#12
    - This work has been punted to the WordPress 6.6 cycle to allow for more time for design work.
    - WordPress/wordpress-develop#5884
 - [ ] WordPress/wordpress-develop#5774
    - this isn't ready yet, as there may be a bug somewhere in it
 - [ ] Explore storing common HTML tag names as a bit field with an "other" category. This could make certain checks in the HTML Processor much cleaner, e.g. `if ( 0 !== $tag_name & (TAG_DIV | TAG_P) )`
 - [ ] Explore storing parsing state as bit fields; same benefit as for tag names.
 - [ ] Exercise 100% test coverage for `step_in_body()`, _*but only by adding high-level usage tests that demonstrate an expected behavior*_.

### Follow-up work to shepherd HTML API usage in Core

 - [ ] WordPress/wordpress-develop#5486
    - seems to be stalled on failing tests
 - [ ] Automattic/jetpack#32700
    - need to address one point of feedback but is otherwise ready

## WordPress-develop Merge Party

 - WordPress/wordpress-develop#5770
 - WordPress/wordpress-develop#5725

The Gutenberg PR WordPress/gutenberg#57241 incorporates a number of proposed changes for Core to give a preview of how they will impact the Gutenberg repo if merged. After briefly breaking the Interactivity API in Gutenberg this should be standard practice before merging new work.

## Awaiting Gutenberg backport

Current backport: WordPress/gutenberg#57022

 - WordPress/WordPress-develop#5535

## Backport Backlog

 - [x] Fix for STYLE and other RAWTEXT elements. (from 6.4)
      - This isn't going to be _fixed_ but instead will be incorporated into the token-scanning PR.

<details><summary>WordPress 6.4</summary>

## Tasks

 - [x] Add `matches_breadcrumbs()` to make it easier to detect location in document in `next_tag()` loop. [WordPress/wordpress-develop#5243]
 - [ ] Continue adding support in `IN BODY` mode. Figure out how much to support.
    - [ ] Support `LI`?
 - [x] Add `has_class()` WordPress/wordpress-develop#5096
 - [ ] Review and refine the error reporting to ensure that a failure to parse is distinguishable from reaching the end of the document.

## Bug fixes
 - [x] Remove all duplicate attributes when removing them, when they exist WordPress/wordpress-develop#4317

## Backport backlog
 - [ ] Fix for `STYLE` and other `RAWTEXT` elements.
 - [x] Revert `while` loop change introduced in styling update.
</details>

**Comment by gziolo** [comment]
What’s the reason it’s a discussion rather than a regular issue? The description is up to date and covers the scope of changes for the WordPress 6.5 release. All other issues like that can be found under the [Epic label](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BType%5D+Epic%22).
